File: web-reference/src/features/clinic-management/utils/staff-list-controls.ts
import { ROLE_LABELS } from '../constants'
import type { BranchRecord, StaffRecord, StaffRole } from '../types'

export type StaffRoleFilter = 'all' | StaffRole
export type StaffBranchFilter = 'all' | string
export type StaffSortKey = 'name-asc' | 'name-desc' | 'role-asc' | 'branch-asc'

export type StaffListControls = {
  search: string
  role: StaffRoleFilter
  branchId: StaffBranchFilter
  sort: StaffSortKey
}

export const STAFF_SORT_OPTIONS = [
  { value: 'name-asc', label: 'Name A–Z' },
  { value: 'name-desc', label: 'Name Z–A' },
  { value: 'role-asc', label: 'Role A–Z' },
  { value: 'branch-asc', label: 'Branch A–Z' },
] as const

export function filterAndSortStaff(
  staff: StaffRecord[],
  branches: BranchRecord[],
  controls: StaffListControls,
): StaffRecord[] {
  let rows = [...staff]
  const q = controls.search.trim().toLowerCase()

  if (q) {
    rows = rows.filter(
      (s) =>
        s.fullName.toLowerCase().includes(q) ||
        s.username.toLowerCase().includes(q) ||
        s.phone.includes(q) ||
        ROLE_LABELS[s.role].toLowerCase().includes(q),
    )
  }

  if (controls.role !== 'all') {
    rows = rows.filter((s) => s.role === controls.role)
  }

  if (controls.branchId !== 'all') {
    rows = rows.filter((s) => s.branchIds.includes(controls.branchId))
  }

  const branchName = (id: string | null) =>
    branches.find((b) => b.id === id)?.name ?? ''

  switch (controls.sort) {
    case 'name-desc':
      rows.sort((a, b) => b.fullName.localeCompare(a.fullName))
      break
    case 'role-asc':
      rows.sort(
        (a, b) =>
          ROLE_LABELS[a.role].localeCompare(ROLE_LABELS[b.role]) ||
          a.fullName.localeCompare(b.fullName),
      )
      break
    case 'branch-asc':
      rows.sort(
        (a, b) =>
          branchName(a.primaryBranchId).localeCompare(branchName(b.primaryBranchId)) ||
          a.fullName.localeCompare(b.fullName),
      )
      break
    default:
      rows.sort((a, b) => a.fullName.localeCompare(b.fullName))
  }

  return rows
}

export const DEFAULT_STAFF_CONTROLS: StaffListControls = {
  search: '',
  role: 'all',
  branchId: 'all',
  sort: 'name-asc',
}

File: frontend/lib/features/clinic-management/presentation/utils/staff_list_controls.dart
import 'package:ai_clinic/features/auth/domain/auth_session.dart';
import 'package:ai_clinic/core/domain/clinic/branch_list_item.dart';
import 'package:ai_clinic/core/domain/clinic/staff_list_item.dart';
import 'package:ai_clinic/features/clinic-management/presentation/constants/clinic_constants.dart';

typedef StaffRoleFilter = StaffRole?;

enum StaffSortKey { nameAsc, nameDesc, roleAsc, branchAsc }

class StaffListControls {
  const StaffListControls({
    this.search = '',
    this.role,
    this.branchId,
    this.sort = StaffSortKey.nameAsc,
  });

  final String search;
  final StaffRoleFilter role;
  final String? branchId;
  final StaffSortKey sort;

  StaffListControls copyWith({
    String? search,
    StaffRoleFilter? role,
    String? branchId,
    StaffSortKey? sort,
    bool clearRole = false,
    bool clearBranchId = false,
  }) {
    return StaffListControls(
      search: search ?? this.search,
      role: clearRole ? null : (role ?? this.role),
      branchId: clearBranchId ? null : (branchId ?? this.branchId),
      sort: sort ?? this.sort,
    );
  }
}

const defaultStaffControls = StaffListControls();

/// Default staff list sort (web `DEFAULT_STAFF_CONTROLS.sort`).
const defaultStaffSort = StaffSortKey.nameAsc;

StaffListControls resetStaffSort(StaffListControls controls) =>
    controls.copyWith(sort: defaultStaffSort);

/// Web `DEFAULT_STAFF_CONTROLS` alias.
// ignore: constant_identifier_names
const DEFAULT_STAFF_CONTROLS = defaultStaffControls;

const staffSortOptions = <({StaffSortKey value, String label})>[
  (value: StaffSortKey.nameAsc, label: 'Name A–Z'),
  (value: StaffSortKey.nameDesc, label: 'Name Z–A'),
  (value: StaffSortKey.roleAsc, label: 'Role A–Z'),
  (value: StaffSortKey.branchAsc, label: 'Branch A–Z'),
];

/// Web `STAFF_SORT_OPTIONS` alias.
// ignore: constant_identifier_names
const STAFF_SORT_OPTIONS = staffSortOptions;

List<StaffListItem> filterAndSortStaff({
  required List<StaffListItem> staff,
  required List<BranchListItem> branches,
  required StaffListControls controls,
}) {
  var rows = [...staff];
  final query = controls.search.trim().toLowerCase();

  if (query.isNotEmpty) {
    rows = rows.where((member) {
      final roleLabel = kRoleLabels[member.role]?.toLowerCase() ?? '';
      return member.fullName.toLowerCase().contains(query) ||
          (member.username?.toLowerCase().contains(query) ?? false) ||
          (member.phone?.contains(query) ?? false) ||
          roleLabel.contains(query);
    }).toList();
  }

  if (controls.role != null) {
    rows = rows.where((member) => member.role == controls.role).toList();
  }

  if (controls.branchId != null && controls.branchId!.isNotEmpty) {
    rows = rows.where((member) => member.isAssignedToBranch(controls.branchId!)).toList();
  }

  String branchNameFor(StaffListItem member) {
    final primary = member.branches.where((branch) => branch.isPrimary).map((branch) => branch.name).firstOrNull;
    if (primary != null) {
      return primary;
    }
    return member.branches.firstOrNull?.name ?? '';
  }

  switch (controls.sort) {
    case StaffSortKey.nameDesc:
      rows.sort((a, b) => b.fullName.toLowerCase().compareTo(a.fullName.toLowerCase()));
    case StaffSortKey.roleAsc:
      rows.sort((a, b) {
        final roleCompare = (kRoleLabels[a.role] ?? '').compareTo(kRoleLabels[b.role] ?? '');
        return roleCompare != 0 ? roleCompare : a.fullName.toLowerCase().compareTo(b.fullName.toLowerCase());
      });
    case StaffSortKey.branchAsc:
      rows.sort((a, b) {
        final branchCompare = branchNameFor(a).compareTo(branchNameFor(b));
        return branchCompare != 0 ? branchCompare : a.fullName.toLowerCase().compareTo(b.fullName.toLowerCase());
      });
    case StaffSortKey.nameAsc:
      rows.sort(StaffListItem.compareByFullName);
  }

  return rows;
}
